### PR TITLE
Replaced ERC20 SVM with ABI SVM documentation 

### PR DESCRIPTION
**Pull request by VGabriel45**

### CODE REVIEW

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md [hunk 1]
```diff
@@ -133,113 +333,181 @@ This function is essential for **validating user operations** against **session
...
 **Operational Flow:**
 
-1. **Decode Session Key Data:** Extracts session key, token address, recipient address, and maximum token amount.
-2. **Validation Checks:** Ensures the destination contract and call value are as required.
-3. **Recipient and Amount Verification:** Compares recipient and transaction amount against session key data.
-4. **Return Session Key:** If all checks pass, returns the session key address.
+The method internally calls **_validateSessionParams** to perform the validation against the session key permissions.
+It passes the necessary parameters to **_validateSessionParams**  and returns the result.
+
+**_validateSessionParams** validates the parameters of the call against the permissions defined by the session key.
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/sessionvalidationmodule.md:498]: defined *for* the session key. not by it.

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md [hunk 2]
```diff
@@ -133,113 +333,181 @@ This function is essential for **validating user operations** against **session
...
+
+**Execution flow in **_validateSessionParams****
+
+- Extracts essential details from _sessionKeyData, such as session key, permitted destination contract, selector, value limit, and rules list length.
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/sessionvalidationmodule.md:502]:
```suggestion
- Extracts essential details from _sessionKeyData, such as session key address, permitted destination contract, permitted method selector, value limit, and rules list length.
```

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md [hunk 3]
```diff
@@ -133,113 +333,181 @@ This function is essential for **validating user operations** against **session
...
+- Checks if the destinationContract matches the permitted destination contract, if the function selector in _funcCallData matches the permitted selector, and if the callValue exceeds the permitted value limit.
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/sessionvalidationmodule.md:503]:
```suggestion
- Checks if the destinationContract matches the permitted destination contract, if the function selector in _funcCallData matches the permitted selector, and if the callValue does not exceed the permitted value limit.
```

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md [hunk 4]
```diff
@@ -133,113 +333,181 @@ This function is essential for **validating user operations** against **session
...
+- Calls **_checkRulesForPermission** to verify if the call data complies with the permission rules.
+
+**Execution flow in **_checkRulesForPermission****
+
+- Iterates through the rules list.
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/sessionvalidationmodule.md:508]:
```suggestion
- Iterates through the list of rules.
```

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md [hunk 5]
```diff
@@ -133,113 +333,181 @@ This function is essential for **validating user operations** against **session
...
+- Parses each rule to determine the offset, condition, and value.
+- Verifies if the call data satisfies each rule condition.
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/sessionvalidationmodule.md:510]:
```suggestion
- Verifies if the call data arguments satisfy conditions defined by the rules.
```

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md [hunk 6]
```diff
@@ -133,113 +333,181 @@ This function is essential for **validating user operations** against **session
...
+**Execution flow in **_validateSessionParams****
+
+- Extracts essential details from _sessionKeyData, such as session key, permitted destination contract, selector, value limit, and rules list length.
+- Checks if the destinationContract matches the permitted destination contract, if the function selector in _funcCallData matches the permitted selector, and if the callValue exceeds the permitted value limit.
+- Calls **_checkRulesForPermission** to verify if the call data complies with the permission rules.
+
+**Execution flow in **_checkRulesForPermission****
+
+- Iterates through the rules list.
+- Parses each rule to determine the offset, condition, and value.
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/sessionvalidationmodule.md:509]: Should we provide more information about `offset`, `condition` and `value`?

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md [hunk 7]
```diff
@@ -133,113 +333,181 @@ This function is essential for **validating user operations** against **session
...
     }
 ```
 
 **Operational Flow:**
 
-1. **Decode Session Key Data:** Extracts session key, token address, recipient address, and maximum token amount.
-2. **Validation Checks:** Ensures the destination contract and call value are as required.
-3. **Recipient and Amount Verification:** Compares recipient and transaction amount against session key data.
-4. **Return Session Key:** If all checks pass, returns the session key address.
+The method internally calls **_validateSessionParams** to perform the validation against the session key permissions.
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/sessionvalidationmodule.md:495]:
```suggestion
The method internally calls **_validateSessionParams** to perform the validation against the session permissions.
```

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md [hunk 8]
```diff
@@ -133,113 +333,181 @@ This function is essential for **validating user operations** against **session
...
 ```
 
 **Execution Steps:**
 
-1. **Match Function Selectors:** Verifies the user operation aligns with specific function selectors.
-2. **Decode Session Key Data:** Extracts essential details like session key, token, recipient, and maximum transaction amount.
-3. **Verify Operation Details:** Checks the operation's token address and call value, and confirms recipient and amount limits.
-4. **Signature Validation:** Utilizes ECDSA to confirm the operation's signature matches the session key.
+1. **Match Function Selectors:** 
+This step ensures that the user operation corresponds to specific function selectors, indicating the type of operation being performed.
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/sessionvalidationmodule.md:384]:
```suggestion
This step ensures that the calldata of a given user operation calls the function with the allowed selectors only.
```

File: docs/tutorials/sessionkeys/createsessionkey.md [hunk 9]
```diff
@@ -233,37 +293,48 @@ const sessionModule = await createSessionKeyManagerModule({
...
-    "0xdA5289fCAAF71d52a80A254da614a192b693e977", // erc20 token address
-    "0x322Af0da66D00be980C7aa006377FCaaEee3BDFD", // receiver address
-    ethers.utils.parseUnits("50".toString(), 6).toHexString(), // 50 usdc amount
-  ],
-);
+const sessionKeyData = await getABISVMSessionKeyData(sessionKeyEOA, {
+    destContract: nftAddress, // destination contract to call
+    functionSelector: functionSelector, // function selector allowed
+    valueLimit: parseEther("0"), // no native value is sent 
+    // In rules, we make sure that referenceValue is equal to recipient
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/createsessionkey.md:302]:
```suggestion
    // This rule ensures that the address we are going to mint NFT to should be equal the reference address (receiver var).
```

File: docs/tutorials/sessionkeys/createsessionkey.md [hunk 10]
```diff
@@ -233,37 +293,48 @@ const sessionModule = await createSessionKeyManagerModule({
...
-    sessionKeyData: sessionKeyData,
+    sessionValidationModule: abiSVMAddress,
+    sessionPublicKey: sessionKeyEOA as `0x${string}`,
+    sessionKeyData: sessionKeyData as `0x${string}`,
   },
 ]);
-console.log("sessionTxData", sessionTxData);
 ```
 
 Now we create the session data itself. We specify how long this should be valid until or how long it is valid after. This should be a unix timestamp to represent the time. Passing 0 on both makes this session never expire, do not do this in production. Next we pass the module address, session key address, and pass the session key data we just created.
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/createsessionkey.md:335]:
```suggestion
Next, we create the data for the transaction that enables the sesson. For every session we need to specify how long it should be valid until and/or how long it is valid after. This should be a unix timestamp to represent the time. Passing 0 on both makes this session never expire, do not do this in production. Next we pass: 
- The Session Validation Module address that is going to validate userOps that leverage this session
- Session key address, that should sign the userOp 
- And the session key data we just created.
```

**Review by filmakarov — CHANGES_REQUESTED** [review]
Hey, added some suggestions

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md (hunk 6 above)
Comment (VGabriel45) [reply docs/tutorials/sessionkeys/sessionvalidationmodule.md:509]: ### Yes. Let me write a bit more about them and I'll ask you to have a look at it again.

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md [hunk 11]
```diff
@@ -509,6 +509,37 @@ It passes the necessary parameters to **_validateSessionParams**  and returns th
...
+ * 0: EQUAL
+ * 1: LESS_THAN_OR_EQUAL
+ * 2: LESS_THAN
+ * 3: GREATER_THAN_OR_EQUAL
+ * 4: GREATER_THAN
+ * 5: NOT_EQUAL
+In our example the condition is 0, this means we check that the receiver of the NFT is EQUAL to what we set it to be.
+
+# Value
+This the the value that we are checking the condition for and that we are getting from the calldata using the offset.
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/sessionvalidationmodule.md:541]: Seems not straightforward enough even for me :) 
I suggest:
`This is the reference value. The actual arg value is decoded from the userOp.callData using the Offset. Then it is compared to the reference value using the Condition.`

File: docs/tutorials/sessionkeys/sessionvalidationmodule.md [hunk 12]
```diff
@@ -509,6 +509,37 @@ It passes the necessary parameters to **_validateSessionParams**  and returns th
...
+So, when dApp is creating a `_sessionKeyData` to enable a session, it should convert every shorter static arg to a 32bytes word to match how it will be actually ABI encoded in the `userOp.callData`.
+
+For the dynamic args, like `bytes`, every 32bytes word of the `calldata` such as offset of the bytes arg, length of the bytes arg, and n-th 32bytes word of the bytes arg can be controlled by a dedicated Rule.
+
+# Offset 
+The offset in the ABI SVM contract helps locate the relevant data within the function call data, it serves as a reference point from which to start reading or extracting specific information required for validation.
+When processing function call data, particularly in low-level languages like Solidity assembly, it's necessary to locate where specific parameters or arguments are stored.
+The offset is used to calculate the starting position within the calldata where the desired data resides.
+Suppose we have a function call with multiple arguments passed as calldata. Each argument occupies a certain number of bytes, and the offset helps determine where each argument begins within the calldata.
+
```
Comment (filmakarov) [inline docs/tutorials/sessionkeys/sessionvalidationmodule.md:529]: I suggest adding a simple example. Like take the method `transfer(address to, uint256 value)` and show what would be offsets for both values

**Review by filmakarov — CHANGES_REQUESTED** [review]
Overall a great explanation, I think it does a great job explaining how things work for those who are interested. I suggested just a couple of changes.

**Review by filmakarov — APPROVED** [review]

**Pull request merged**